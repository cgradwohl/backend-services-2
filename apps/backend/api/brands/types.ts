import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiBrandsGetResponse,
  IApiBrandsListResponse,
  IApiBrandsPostRequest,
  IApiBrandsPostResponse,
  IApiBrandsPutRequest,
  IApiBrandsPutResponse,
} from "~/types.public";

export type ApiBrandsResponse =
  | IApiBrandsGetResponse
  | IApiBrandsListResponse
  | IApiBrandsPostResponse
  | IApiBrandsPutResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsGetResponse;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsListResponse;
}>;

export type PostRequestBody = IApiBrandsPostRequest;

export type PostFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsPostResponse;
}>;

export type PutRequestBody = IApiBrandsPutRequest;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsPutResponse;
}>;

export type RemoveFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204;
}>;
