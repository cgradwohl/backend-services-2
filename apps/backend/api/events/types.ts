import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiEventsGetResponse,
  IApiEventsListResponse,
  IApiEventsPutRequest,
  IApiEventsPutResponse,
} from "~/types.public";

export type ApiEventsResponse = IApiEventsGetResponse | IApiEventsListResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsGetResponse;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsListResponse;
}>;

export type PutRequestBody = IApiEventsPutRequest;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiEventsPutResponse;
}>;

export type RemoveFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204;
}>;
