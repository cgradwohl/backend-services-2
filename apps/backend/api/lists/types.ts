import { ApiRequestContext, Response } from "~/lib/lambda-response";

import {
  IApiDeleteListItemResponse,
  IApiGetListItemResponse,
  IApiGetListItemsResponse,
  IApiGetListItemSubscriptionsResponse,
  IApiListItemSubscribeResponse,
  IApiListItemUnsubscribeResponse,
  IApiPutListItemResponse,
  IApiRestoreListItemResponse,
} from "~/types.public";

export type HttpEventHandler<T extends Response<any>> = (
  context: ApiRequestContext
) => Promise<T>;

export type ListApiResponse =
  | IApiDeleteListItemResponse
  | IApiGetListItemResponse
  | IApiGetListItemsResponse
  | IApiGetListItemSubscriptionsResponse
  | IApiListItemSubscribeResponse
  | IApiListItemUnsubscribeResponse
  | IApiPutListItemResponse
  | IApiRestoreListItemResponse;
