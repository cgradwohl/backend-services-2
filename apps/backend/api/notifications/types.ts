import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiNotificationGetResponse,
  IApiNotificationGetSubmissionChecksResponse,
  IApiNotificationListResponse,
  IApiNotificationPutSubmissionChecksResponse,
} from "~/types.public";

export type ApiNotificationResponse =
  | IApiNotificationGetResponse
  | IApiNotificationListResponse
  | IApiNotificationGetSubmissionChecksResponse
  | IApiNotificationPutSubmissionChecksResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationGetResponse;
}>;

export type PostFn = (
  context: ApiRequestContext
) => Promise<{
  status: number;
}>;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  status: number;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationListResponse;
}>;

export type GetSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationGetSubmissionChecksResponse;
}>;

export type PutSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiNotificationPutSubmissionChecksResponse;
}>;

export type DeleteSubmissionChecksFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204 | 409;
}>;
