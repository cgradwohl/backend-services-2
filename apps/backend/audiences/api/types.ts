import { ApiRequestContext } from "~/lib/lambda-response";
import { FilterConfig } from "~/audiences/stores/dynamo/types";

export type Resource =
  | "/audiences"
  | "/audiences/{audience_id}"
  | "/audiences/{audience_id}/members"
  | "/audiences/{audience_id}/members/{member_id}";

export type MemberResource = "/members/{member_id}/audiences";

export type BaseAudienceResponse = {};

export interface AudienceHandler<T extends BaseAudienceResponse> {
  (context: ApiRequestContext): Promise<T>;
}

export type Audience = {
  created_at: string;
  description: string;
  id: string;
  name: string;
  // TODO: https://linear.app/trycourier/issue/C-5533/rule-engine-ux-for-allowing-people-to-create-audiences-rule
  // rules will never be null, but until we have a UI for creating audiences, we need to allow it to be null
  filter: FilterConfig | null;
  updated_at: string;
};

export type AudienceMember = {
  added_at: string;
  audience_id: string;
  audience_version: number;
  member_id: string;
  reason: string;
};

export interface AudienceListResponse extends BaseAudienceResponse {
  body: {
    paging: {
      cursor: string;
      more: boolean;
    };
    items: Audience[];
  };
}

export interface AudienceDeleteResponse extends BaseAudienceResponse {
  status: 204;
}

export interface AudienceGetResponse extends BaseAudienceResponse {
  status: 200;
  body: Audience;
}

export interface AudienceMemberGetResponse extends BaseAudienceResponse {
  status: 200;
  body: AudienceMember;
}

export interface AudienceMemberListResponse extends BaseAudienceResponse {
  status: 200;
  body: {
    items: AudienceMember[];
    paging: {
      cursor: string;
      more: boolean;
    };
  };
}

export interface AudiencePutResponse extends BaseAudienceResponse {
  status: 200;
  body: {
    audience: Audience;
  };
}

export interface AudienceMemberPutResponse extends BaseAudienceResponse {
  status: 200;
  body: {
    audience_member: AudienceMember;
  };
}

export type Method = "get" | "put" | "delete";

export type UserMethod = "get";
