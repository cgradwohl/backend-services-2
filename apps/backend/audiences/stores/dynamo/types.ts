import { Operator } from "~/audiences/lib/audience-rule-engine";
export interface IDDBAudienceKeys {
  gsi1pk: string;
  pk: string;
  shardId?: number;
  gsi2pk?: string;
}

type Without<T, U> = { [P in Exclude<keyof T, keyof U>]?: never };
type XOR<T, U> = T | U extends object
  ? (Without<T, U> & U) | (Without<U, T> & T)
  : T | U;

interface BaseFilterConfig {
  operator: Operator;
}

export interface SingleFilterConfig extends BaseFilterConfig {
  path: string;
  value: string | boolean;
}

interface NestedFilterConfig extends BaseFilterConfig {
  filters: FilterConfig[];
}

export type FilterConfig = XOR<SingleFilterConfig, NestedFilterConfig>;

export interface IDDBAudience extends IDDBAudienceKeys {
  audienceId: string;
  createdAt: string;
  description?: string;
  lastSendAt?: string;
  memberCount?: number;
  name?: string;
  updatedAt: string;
  version: number;
  filter: FilterConfig;
  workspaceId: string;
}

export interface IDDBAudienceMember extends IDDBAudienceKeys {
  addedAt: string;
  audienceId: string;
  audienceVersion: number;
  userId: string;
  reason: string;
  workspaceId: string;
}

export interface IDDBAudienceCalculation extends IDDBAudienceKeys {
  audienceId: string;
  lastUpdatedAt: string;
  result: "calculating" | "calculated" | "error";
  totalUsers: number;
  totalUsersFiltered: number;
  userCount: number;
  workspaceId: string;
}
