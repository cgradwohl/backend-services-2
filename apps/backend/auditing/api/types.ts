import { ApiRequestContext } from "~/lib/lambda-response";
import { IAuditEventApiOutput } from "../services/search";

export interface IApiAuditEventsListResponse {
  paging: {
    cursor?: string;
    more: boolean;
  };
  results: IAuditEventApiOutput[];
}

export type ListFn = (context: ApiRequestContext) => Promise<{
  body: IApiAuditEventsListResponse;
}>;

export type GetFn = (context: ApiRequestContext) => Promise<{
  body: IAuditEventApiOutput;
}>;
