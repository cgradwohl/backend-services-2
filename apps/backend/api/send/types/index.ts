import {
  IBrandColors,
  IBrandSettingsEmail,
  IBrandSettingsInApp,
  IBrandSnippets,
} from "~/lib/brands/types";
import { IBrand } from "~/types.api";
import { IProfilePreferences } from "~/types.public";
import { ElementalContent } from "./courier-elemental";

export * from "./courier-elemental";

export interface MessageData extends Record<string, any> {}

type InvalidListRecipient = {
  user_id: string;
  list_pattern: string;
};

type ListRecipientType = Record<string, unknown> & {
  [key in keyof InvalidListRecipient]?: never;
};
export interface ListRecipient extends ListRecipientType {
  list_id?: string;
  data?: MessageData;
}

export interface AudienceRecipient {
  audience_id: string;
  data?: MessageData;
}

type InvalidListPatternRecipient = {
  user_id: string;
  list_id: string;
};

type ListPatternRecipientType = Record<string, unknown> & {
  [key in keyof InvalidListPatternRecipient]?: never;
};
export interface ListPatternRecipient extends ListPatternRecipientType {
  list_pattern?: string;
  data?: MessageData;
}

type InvalidUserRecipient = {
  list_id: string;
  list_pattern: string;
};

type UserRecipientType = Record<string, unknown> & {
  [key in keyof InvalidUserRecipient]?: never;
};

export interface UserRecipient extends UserRecipientType {
  data?: MessageData;
  email?: string;
  locale?: string;
  user_id?: string;
  phone_number?: string;
  preferences?: IProfilePreferences;
}

export type Recipient =
  | AudienceRecipient
  | ListRecipient
  | ListPatternRecipient
  | UserRecipient;

export type MessageRecipient = Recipient | Recipient[];

export interface ElementalContentSugar {
  body?: string;
  title?: string;
}

export interface Timeout {
  /** I recommend deprecating the record format, makes overrides a bit confusing */
  provider?: number | { [provider: string]: number };
  /** I recommend deprecating the record format, makes overrides a bit confusing */
  channel?: number | { [provider: string]: number };
  message?: number;
}

/** @deprecated */
export interface TimeoutDate<T> {
  message: T;
  channel?: T;
  channels?: {
    [channel: string]: ChannelTimeout<T>;
  };
  provider?: T;
  providers?: { [provider: string]: T };
}

/** @deprecated Timeout specified as an epoch date (seconds after unix epoch) */
export interface TimeoutDateEpochSeconds extends TimeoutDate<number> {}

/** @deprecated Timeout specified as ISO 8601 date string  */
export interface TimeoutDateIso8601 extends TimeoutDate<string> {}

export interface ChannelTimeout<T> {
  channel?: T;
  provider?: T;
}

export type Content = ElementalContentSugar | ElementalContent;

export interface MessageDelayDuration {
  duration: string; // friendly ("1 hour"), milliseconds, ISO-8601 duration
  until?: never;
}

export interface MessageDelayUntil {
  duration?: never;
  until: string; // friendly ("Thursday at 10am"), Unix Epoch, ISO-8601 date-stamp w/ TZ
}

export type MessageDelay = MessageDelayDuration | MessageDelayUntil;

export interface MessageBrandV2 {
  version: "2022-05-17";
  colors?: {
    primary?: string;
    secondary?: string;
    tertiary?: string;
  };
  logo?: {
    href?: string;
    image: string;
  };
  locales?: {
    [key: string]: Omit<MessageBrand, "locales" | "version">;
  };
}

export type MessageBrandV1 = Partial<Omit<IBrand, "version">> & {
  settings: IBrand["settings"];
  version?: "2020-06-19T18:51:36.083Z" | "2020-06-19";
};

export type MessageBrand = MessageBrandV1 | MessageBrandV2;

export interface BaseMessage {
  brand_id?: string;
  brand?: MessageBrand;
  channels?: MessageChannels;
  data?: MessageData;
  delay?: MessageDelay;
  metadata?: MessageMetadata;
  providers?: MessageProviders;
  routing?: MessageRouting;
  timeout?: Timeout;
  to: MessageRecipient;
}

interface TrackingOverride {
  open: boolean;
}
export interface MessageChannelEmailOverride {
  attachments?: Record<string, any>[];
  bcc?: string;
  // content?: Content; TODO: https://linear.app/trycourier/issue/C-4462/add-content-support-to-channel-overrides
  brand?: {
    snippets?: IBrandSnippets;
    settings?: {
      colors?: IBrandColors;
      email?: IBrandSettingsEmail;
    };
  };
  cc?: string;
  from?: string;
  html?: string;
  reply_to?: string;
  subject?: string;
  text?: string;
  tracking?: TrackingOverride;
}

export interface MessageChannelPushOverride {
  body?: string;
  brand?: {
    snippets?: IBrandSnippets;
    settings?: {
      colors?: IBrandColors;
      inapp?: IBrandSettingsInApp;
    };
  };
  // content?: Content; TODO: https://linear.app/trycourier/issue/C-4462/add-content-support-to-channel-overrides
  click_action?: string;
  data?: Record<string, any>;
  icon?: string;
  reply_to?: undefined;
  title?: string;
}

export interface MessageProviderConfig {
  override?: Record<string, any>;
  if?: string;
  timeout?: number;
  metadata?: {
    utm?: UTM;
  };
}

export interface MessageProviders {
  [provider: string]: MessageProviderConfig;
}

export interface MessageChannelConfig {
  brand_id?: string; // TODO: pick this up in prepare (like override)?
  providers?: string[];
  routing_method?: "all" | "single";
  if?: string;
  timeout?: number | ChannelTimeout<number>;
  metadata?: {
    utm?: UTM;
  };
  override?: MessageChannelEmailOverride | MessageChannelPushOverride;
}
export interface MessageChannels {
  [channel: string]: MessageChannelConfig;
}

export type MessageRoutingMethod = "all" | "single";
export type MessageRoutingChannel = string | MessageRouting;
export interface MessageRouting {
  method: MessageRoutingMethod;
  channels: MessageRoutingChannel[];
}

export interface UTM {
  source?: string;
  medium?: string;
  campaign?: string;
  term?: string;
  content?: string;
}

export interface UTMMap {
  message?: UTM;
  channels?: {
    [channel: string]: UTM;
  };
  providers?: {
    [provider: string]: UTM;
  };
}

export interface MessageMetadata {
  event?: string;
  tags?: string[];
  utm?: UTM;
  trace_id?: string;
}

export interface ContentMessage extends BaseMessage {
  content: Content;
}

export interface TemplateMessage extends BaseMessage {
  template: string;
}

export type Message = ContentMessage | TemplateMessage;

export enum SequenceActions {
  cancel = "cancel",
  emit = "emit",
  fetchData = "fetch-data",
  send = "send",
}

export interface SendSequenceAction {
  action: SequenceActions;
  message: Message;
}

export type Sequence = Array<SendSequenceAction>;
// | EmitSequenceAction
// | FetchDataSequenceAction
// | CancelSequenceAction>;

export interface MessageRequest {
  message: Message;
  sequence?: never;
}

export interface SequenceRequest {
  message?: never;
  sequence: Sequence;
}

export type RequestV2 = SequenceRequest | MessageRequest;
