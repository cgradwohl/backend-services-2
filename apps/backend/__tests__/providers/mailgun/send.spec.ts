import axios from "axios";
import createLinkHandler from "~/lib/link-handler";
import createVariableHandler from "~/lib/variable-handler";
import { ProviderResponseError } from "~/providers/errors";
import * as formHelper from "~/providers/mailgun/create-form-data";
import sendHandler from "~/providers/mailgun/send";
import { DeliveryHandlerParams } from "~/providers/types";

jest.mock("axios");

const axiosPostSpy = jest.spyOn(axios, "post");

const createFormDataSpy = jest.spyOn(formHelper, "createFormData");

const getDeliveryParams = (
  testConfig?: any,
  body: any = { profile: {} }
): DeliveryHandlerParams => {
  const variableData = {
    event: "",
    recipient: "",
    ...body,
  };

  const variableHandler = createVariableHandler({ value: variableData });
  const linkHandler = createLinkHandler({});

  return {
    config: {
      ...testConfig,
      provider: "",
    },
    linkHandler,
    profile: body.profile,
    variableData,
    variableHandler,
  };
};

const config = {
  apiKey: "mockApiKey",
  domain: "mockDomain",
};
const profile = {
  email: "to@mail.com",
};

const params = getDeliveryParams(config, {
  profile,
});

const template = {
  bcc: "bcc@mail.com",
  cc: "cc@mail.com",
  from: "from@mail.com",
  html: "",
  replyTo: "replyTo@mail.com",
  subject: "",
  text: "",
};
const expectedMailgunResponse = {
  data: {
    id: "<20211013230936.1.C913CB59DE255FB1@sandboxea24b22d01174cfc9c9ea70959d6b274.mailgun.org>",
    message: "Queued. Thank you.",
  },
};

describe("mailgun basic send", () => {
  beforeEach(() => {
    axiosPostSpy.mockResolvedValue(expectedMailgunResponse);
  });
  afterAll(() => {
    createFormDataSpy.mockReset();
  });

  it("all with correct keys", async () => {
    const response = await sendHandler(params, template);
    expect(response).toMatchObject(expectedMailgunResponse.data);
    expect(axiosPostSpy.mock.calls[0][0]).toMatch(
      "https://api.mailgun.net/v3/mockDomain/messages"
    );
    const { replyTo, ...rest } = template;
    expect(createFormDataSpy).toHaveBeenCalledWith({
      ...rest,
      "h:Reply-To": "replyTo@mail.com",
      to: "to@mail.com",
    });
  });
});

describe("mailgun overrides send", () => {
  beforeAll(() => {
    axiosPostSpy.mockResolvedValue(expectedMailgunResponse);

    params.override = {
      body: {
        "o:tag": "notifications",
      },
      config: {
        apiKey: "newApiKey",
        host: "newHost",
      },
    };
  });

  afterAll(() => {
    createFormDataSpy.mockReset();
    params.override = undefined;
  });

  it("send should respect overrides", async () => {
    const response = await sendHandler(params, template);

    expect(axiosPostSpy.mock.calls[0][0]).toMatch(
      "https://api.mailgun.net/v3/mockDomain/messages"
    );

    expect(response).toMatchObject(expectedMailgunResponse.data);

    expect(Object.keys(axiosPostSpy.mock.calls[0][2])).toMatchSnapshot(
      "should match headers"
    );

    expect(createFormDataSpy.mock.calls[0][0]).toMatchSnapshot(
      "should match form data"
    );

    expect(
      axiosPostSpy.mock.calls[0][2].headers["content-type"].startsWith(
        "multipart/form-data"
      )
    ).toBeTruthy();
  });
});

describe("mailgun attachents send", () => {
  beforeEach(() => {
    axiosPostSpy.mockResolvedValue(expectedMailgunResponse);
  });

  it("mailgun should get attachments", async () => {
    params.override = {
      attachments: [
        {
          contentType: "application/pdf",
          data: "Q29uZ3JhdHVsYXRpb25zLCB5b3UgY2FuIGJhc2U2NCBkZWNvZGUh",
          filename: "billing.pdf",
        },
      ],
    };
    await sendHandler(params, template);

    const expectedFormDataInput = [
      "from",
      "h:Reply-To",
      "html",
      "subject",
      "text",
      "to",
      "attachment",
      "cc",
      "bcc",
    ];

    const actualFormDataInput = Object.keys(createFormDataSpy.mock.calls[0][0]);
    expect(actualFormDataInput).toMatchObject(expectedFormDataInput);
  });
});

describe("mailgun incorrect domain errors", () => {
  beforeEach(() => {
    axiosPostSpy.mockResolvedValue({
      data: "Mailgun Magnificent API",
    });
  });

  it("should throw providerResponseError", async () => {
    await expect(sendHandler(params, template)).rejects.toThrow(
      ProviderResponseError
    );
  });
});
