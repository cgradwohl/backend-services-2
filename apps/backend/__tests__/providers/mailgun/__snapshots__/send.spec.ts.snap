// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mailgun overrides send send should respect overrides: should match form data 1`] = `
Object {
  "bcc": "bcc@mail.com",
  "cc": "cc@mail.com",
  "from": "from@mail.com",
  "h:Reply-To": "replyTo@mail.com",
  "html": "",
  "o:tag": "notifications",
  "subject": "",
  "text": "",
  "to": "to@mail.com",
}
`;

exports[`mailgun overrides send send should respect overrides: should match headers 1`] = `
Array [
  "auth",
  "headers",
  "timeout",
  "timeoutErrorMessage",
]
`;
